File: server/src/services/member_service.ts
import { PrismaClient } from '@prisma/client';
import { BusinessError } from '@/middlewares/error_middleware';
import { computeUserRights } from '@/utils/user_rights';

const prisma = new PrismaClient();

const DEFAULT_PLANS: Record<string, { name: string; price: number; days: number; memberLevel: number }> = {
    monthly: { name: '月卡', price: 9.9, days: 30, memberLevel: 1 },
    quarterly: { name: '季卡', price: 25.0, days: 90, memberLevel: 1 },
    yearly: { name: '包年', price: 88.0, days: 365, memberLevel: 2 },
};

async function getPlans() {
    // 未来可从 SystemConfig 覆盖价格，当前返回默认值
    return DEFAULT_PLANS;
}

async function getInfo(userId: number) {
    const user = await prisma.user.findUnique({
        where: { id: userId },
        include: { member: true },
    });
    if (!user) return null;

    const rights = computeUserRights({
        openId: user.openId,
        status: user.status,
        memberLevel: user.memberLevel,
        memberExpireAt: user.memberExpireAt,
        trialExpireAt: user.trialExpireAt,
    });

    return {
        id: user.id,
        nickname: user.nickname,
        avatar: user.avatar,
        memberLevel: rights.memberLevel,
        isPaidMember: rights.isPaidMember,
        isTrialActive: rights.isTrialActive,
        trialExpireAt: rights.trialExpireAt?.toISOString(),
        memberExpireDate: user.memberExpireAt?.toISOString(),
        planType: user.member?.planType,
        canUseFeature: rights.canUseFeature,
    };
}

async function activate(userId: number, planType: string): Promise<void> {
    const plans = await getPlans();
    const plan = plans[planType];
    if (!plan) throw new BusinessError(400, '无效的套餐类型');

    const now = new Date();
    const future = new Date(now.getTime() + plan.days * 24 * 60 * 60 * 1000);

    const user = await prisma.user.findUnique({ where: { id: userId } });
    if (!user) throw new BusinessError(404, '用户不存在');

    // 计算新的 memberExpireAt：若当前会员未过期则续期
    let newExpireAt = future;
    if (user.memberExpireAt && user.memberExpireAt.getTime() > now.getTime()) {
        newExpireAt = new Date(user.memberExpireAt.getTime() + plan.days * 24 * 60 * 60 * 1000);
    }

    await prisma.$transaction([
        prisma.user.update({
            where: { id: userId },
            data: {
                memberLevel: plan.memberLevel,
                memberExpireAt: newExpireAt,
            },
        }),
        prisma.member.upsert({
            where: { userId },
            create: {
                userId,
                planType,
                startDate: now,
                expireDate: newExpireAt,
            },
            update: {
                planType,
                startDate: now,
                expireDate: newExpireAt,
                status: 1,
            },
        }),
    ]);
}

function generateOrderNo(): string {
    return 'SK' + Date.now() + Math.floor(Math.random() * 1000);
}

async function createOrder(
    userId: number,
    planType: string,
    payChannel = 'wechat',
): Promise<{ id: number; orderNo: string; planType: string; amount: number; payStatus: number; createdAt: string }> {
    const plans = await getPlans();
    const plan = plans[planType];
    if (!plan) throw new BusinessError(400, '无效的套餐类型');

    const orderNo = generateOrderNo();
    const order = await prisma.order.create({
        data: {
            userId,
            orderNo,
            planType,
            amount: plan.price,
            payChannel,
            payStatus: 0,
        },
    });

    return {
        id: order.id,
        orderNo: order.orderNo,
        planType: order.planType,
        amount: order.amount,
        payStatus: order.payStatus,
        createdAt: order.createdAt.toISOString(),
    };
}

async function markPaid(orderNo: string, payChannel?: string): Promise<void> {
    const order = await prisma.order.findUnique({ where: { orderNo } });
    if (!order) throw new BusinessError(404, '订单不存在');
    if (order.payStatus === 1) return;

    await prisma.order.update({
        where: { orderNo },
        data: {
            payStatus: 1,
            payTime: new Date(),
            ...(payChannel ? { payChannel } : {}),
        },
    });

    await activate(order.userId, order.planType);
}

export { getInfo, getPlans, activate, createOrder, markPaid };
export default { getInfo, getPlans, activate, createOrder, markPaid };

File: server/app/services/member_service.py
"""会员服务（对应 TS member_service.ts）"""
import random
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

from ..deps import BusinessError
from ..models import Member, Order, User
from ..utils.user_rights import compute_user_rights

DEFAULT_PLANS = {
    "monthly": {"name": "月卡", "price": 9.9, "days": 30, "memberLevel": 1},
    "quarterly": {"name": "季卡", "price": 25.0, "days": 90, "memberLevel": 1},
    "yearly": {"name": "包年", "price": 88.0, "days": 365, "memberLevel": 2},
}


def getPlans() -> dict:
    # 未来可从 SystemConfig 覆盖价格，当前返回默认值
    return DEFAULT_PLANS


def _findMember(db: Session, userId: int) -> Member | None:
    return db.query(Member).filter(Member.userId == userId).first()


def getInfo(db: Session, userId: int) -> dict | None:
    user = db.query(User).filter(User.id == userId).first()
    if not user:
        return None

    rights = compute_user_rights(
        {
            "openId": user.openId,
            "status": user.status,
            "memberLevel": user.memberLevel,
            "memberExpireAt": user.memberExpireAt,
            "trialExpireAt": user.trialExpireAt,
        }
    )

    member = _findMember(db, userId)
    trial_expire_at = rights.get("trialExpireAt")
    return {
        "id": user.id,
        "nickname": user.nickname,
        "avatar": user.avatar,
        "memberLevel": rights["memberLevel"],
        "isPaidMember": rights["isPaidMember"],
        "isTrialActive": rights["isTrialActive"],
        "trialExpireAt": trial_expire_at.isoformat() if trial_expire_at else None,
        "memberExpireDate": user.memberExpireAt.isoformat() if user.memberExpireAt else None,
        "planType": member.planType if member else None,
        "canUseFeature": rights["canUseFeature"],
    }


def activate(db: Session, userId: int, planType: str) -> None:
    plans = getPlans()
    plan = plans.get(planType)
    if not plan:
        raise BusinessError(400, "无效的套餐类型")

    now = datetime.now()
    future = now + timedelta(days=plan["days"])

    user = db.query(User).filter(User.id == userId).first()
    if not user:
        raise BusinessError(404, "用户不存在")

    # 计算新的 memberExpireAt：若当前会员未过期则续期
    new_expire_at = future
    if user.memberExpireAt and user.memberExpireAt > now:
        new_expire_at = user.memberExpireAt + timedelta(days=plan["days"])

    try:
        user.memberLevel = plan["memberLevel"]
        user.memberExpireAt = new_expire_at
        user.updatedAt = now

        member = _findMember(db, userId)
        if member:
            member.planType = planType
            member.startDate = now
            member.expireDate = new_expire_at
            member.status = 1
            member.updatedAt = now
        else:
            db.add(
                Member(
                    userId=userId,
                    planType=planType,
                    startDate=now,
                    expireDate=new_expire_at,
                    status=1,
                )
            )
        db.commit()
    except Exception:
        db.rollback()
        raise


def generateOrderNo() -> str:
    return "SK" + str(int(datetime.now().timestamp() * 1000)) + str(random.randint(0, 999))


def createOrder(
    db: Session, userId: int, planType: str, payChannel: str = "wechat"
) -> dict:
    plans = getPlans()
    plan = plans.get(planType)
    if not plan:
        raise BusinessError(400, "无效的套餐类型")

    orderNo = generateOrderNo()
    order = Order(
        userId=userId,
        orderNo=orderNo,
        planType=planType,
        amount=plan["price"],
        payChannel=payChannel,
        payStatus=0,
    )
    db.add(order)
    db.commit()
    db.refresh(order)

    return {
        "id": order.id,
        "orderNo": order.orderNo,
        "planType": order.planType,
        "amount": order.amount,
        "payStatus": order.payStatus,
        "createdAt": order.createdAt.isoformat() if order.createdAt else None,
    }


def markPaid(db: Session, orderNo: str, payChannel: str | None = None) -> None:
    order = db.query(Order).filter(Order.orderNo == orderNo).first()
    if not order:
        raise BusinessError(404, "订单不存在")
    if order.payStatus == 1:
        return

    update_data = {"payStatus": 1, "payTime": datetime.now()}
    if payChannel:
        update_data["payChannel"] = payChannel

    try:
        db.query(Order).filter(Order.orderNo == orderNo).update(update_data)
        db.commit()
    except Exception:
        db.rollback()
        raise

    activate(db, order.userId, order.planType)


def checkIsMember(db: Session, userId: int) -> bool:
    """判断用户是否为有效付费会员（custom_quote_service 依赖）"""
    user = db.query(User).filter(User.id == userId).first()
    if not user:
        return False
    rights = compute_user_rights(
        {
            "openId": user.openId,
            "status": user.status,
            "userType": user.userType,
            "memberLevel": user.memberLevel,
            "memberExpireAt": user.memberExpireAt,
            "trialExpireAt": user.trialExpireAt,
        }
    )
    return bool(rights["isPaidMember"])
